File: backend_js/src/graphbuilder.ts
/**
 * Graph Builder for HR Agent
 * Maps from Python graphbuilder.py
 */

import { StateGraph, START, END, MemorySaver } from "@langchain/langgraph";
import { toolsCondition } from "@langchain/langgraph/prebuilt";
import { ChatGroq } from "@langchain/groq";
import type { ToolNode } from "@langchain/langgraph/prebuilt";
import { StateAnnotation } from "./state.js";
import { HRNode } from "./nodes.js";

/**
 * GraphBuilder class
 * Maps from Python GraphBuilder class
 */
export class GraphBuilder {
  private llm: ChatGroq;
  private toolNode: ToolNode | null;
  private graph: StateGraph<any>;

  constructor(llm: ChatGroq, toolNode: ToolNode | null) {
    this.llm = llm;
    this.toolNode = toolNode;
    this.graph = new StateGraph(StateAnnotation);
  }

  /**
   * Builds the graph used to execute HR queries
   * Maps from Python build_graph method
   */
  buildGraph() {
    // Initialize the HR node
    const hrNode = new HRNode(this.llm);

    // Use provided tool_node (should be passed from app startup)
    if (this.toolNode === null) {
      throw new Error("tool_node must be provided to GraphBuilder");
    }

    // Build the graph using method chaining
    const compiledGraph = new StateGraph(StateAnnotation)
      // Add the nodes to the graph
      .addNode("hr_node", (state, config) => hrNode.execute(state, config))
      .addNode("tools", this.toolNode)
      // Add the edges to the graph
      .addEdge(START, "hr_node")
      // Conditional edge: if tools are needed, route to tools, otherwise END
      .addConditionalEdges("hr_node", toolsCondition, {
        tools: "tools",
        __end__: END,
      })
      // After tools execute, route back to hr_node
      .addEdge("tools", "hr_node")
      // Compile with memory checkpointing
      .compile({
        checkpointer: new MemorySaver(),
      });

    return compiledGraph;
  }
}

// Export the graph for langgraph.json
// This will be built at runtime in app.ts
export let graph: ReturnType<GraphBuilder["buildGraph"]> | null = null;

File: backend/src/agents/db_agent/graphbuilder.py
from langgraph.graph import START, StateGraph, END
from langgraph.prebuilt import tools_condition
from src.agents.db_agent.state import State
from src.agents.db_agent.nodes import DB_Node
from langgraph.checkpoint.memory import MemorySaver


class DB_Agent_GraphBuilder:
    """
    This class builds the graph used to execute DB operations.
    """

    def __init__(self, llm, tool_node):
        self.llm = llm
        self.tool_node = tool_node
        self.graph = StateGraph(State)

    
    def build_graph(self):

        db_node = DB_Node(self.llm)

        self.graph.add_node("extract_text", db_node.read_file)
        self.graph.add_node("get_employee_uuid", db_node.get_employee_uuid)
        self.graph.add_node("generate_summary", db_node.generate_summary)
        self.graph.add_node("upload_to_storage", db_node.upload_to_storage)
        self.graph.add_node("update_table", db_node.update_table)
        self.graph.add_node("tools", self.tool_node)

        self.graph.add_edge(START, "extract_text")
        self.graph.add_edge("extract_text", "upload_to_storage")
        self.graph.add_edge("upload_to_storage", "generate_summary")
        self.graph.add_edge("generate_summary", "get_employee_uuid")


        self.graph.add_conditional_edges(
            "get_employee_uuid",
            tools_condition,
            {
                "tools": "tools",
                END: "update_table"
            }
        )

        self.graph.add_edge("tools", "get_employee_uuid")

        self.graph.add_edge("update_table", END)

        return self.graph.compile()